### ISSUE TITLE
[Feature Request] Export TS types, interfaces

### ISSUE THREAD

**Issue by websitevirtuoso** [vuetifyjs/vuetify#16680]
### Problem to solve
Let's consider next struct
```
const headers: DataTableHeader[] = [
  { title: '#', key: 'id' },
  { title: t('messages.name'), key: 'name' },
  { title: t('messages.lat'), key: 'lat' },
  { title: t('messages.lng'), key: 'lng' },
  { title: t('messages.state'), key: 'state', value: 'state.name', sortable: false },
  { title: t('messages.country'), key: 'country', value: 'state.country.name', sortable: false },
]
if(can('upsert', 'city')) {
  headers.push({ title: t('messages.actions'), key: 'action', sortable: false, align: 'end' })
}
```

We as developers don't have access to defined types, interfaces as a result in my project I have to duplicate the same TS types

### Proposed solution
Just add export to types, interfaces that you already defined

**Comment by wuxin0011** [comment]
good!

**Comment by wuxin0011** [comment]
I also encountered the same problem

**Comment by websitevirtuoso** [comment]
I have no idea about all types. Because I used only types from datatable. I had to copy this types into my own code and use them with "todo commend". I hope that they will create index.d.ts for us with exported types and interfaces

**Comment by felicienfrancois** [comment]
@joel-wenzel a simpler variant without typescript error
```
export type VDataTableHeaderProps = Extract<
  VDataTable['$props']['headers'][number],
  {key: string}
>
```

**Comment by nowaysgit** [comment]
waiting...

**Comment by nowaysgit** [comment]
> @joel-wenzelболее простой вариант без ошибок машинописного текста
> 
> ```ts
> import { VDataTable } from 'vuetify/labs/VDataTable';
> export type VDataTableHeaderProps = Extract<VDataTable['$props']['headers'][number], {key: string}>
> ```

very sad situation
```
error TS2537: Type 'DeepReadonly<DataTableHeader[] | DataTableHeader[][]> | undefined' has no matching index signature for type 'number'.

145   export type VDataTableHeaderProps = Extract<VDataTable["$props"]["headers"][number], { key: string }>;
```

**Comment by felicienfrancois** [comment]
@nowaysgit 

```typescript
export type VDataTableHeaderProps = Extract<Extract<VDataTable["$props"]["headers"], Readonly<Array>>[number], {key: string}>
```

**Comment by MatthewAry** [comment]
See the following for references and examples:
- https://vuejs.org/guide/typescript/composition-api.html
- https://www.radix-vue.com/guides/styling.html#extending-a-primitive

```
<script setup lang="ts">
import { VCard, type VCardProps } from "vuetify/components/VCard";

interface Props extends VCardProps {
  foo: string;
}

const props = defineProps<Props>();
</script>

<template>
  <VCard v-bind="props" />
</template>
```

**Comment by websitevirtuoso** [comment]
Yes I can confirm that rignt now is much better with exports types and interfaces.

For example:
my code before:
```
import { DataTableHeader } from '@/types/vuetify'

const headers: DataTableHeader[] = [
  { title: '#', key: 'id', sortable: false },
  { title: t('messages.name'), key: 'name', sortable: false },
  { title: t('messages.description'), key: 'description', sortable: false },
]

if (can('update', 'listing_term')) {
  headers.push({ title: t('messages.actions'), key: 'action', sortable: false, align: 'end' })
}
```

my file @/types/vuetify with all duplicated types from vuetify
code after refactoring vuetify and using their types

```
import { VDataTableColumn } from 'vuetify/lib/labs/VDataTable/VDataTableColumn'

const headers: VDataTableColumn[] = [
  { title: '#', key: 'id', sortable: false },
  { title: t('messages.name'), key: 'name', sortable: false },
  { title: t('messages.description'), key: 'description', sortable: false },
]

if (can('update', 'listing_term')) {
  headers.push({ title: t('messages.actions'), key: 'action', sortable: false, align: 'end' })
}
```

So can I close this issue?

**Comment by MatthewAry** [comment]
@websitevirtuoso I don't think so. It seems like the team has expanded the scope of this issue such that it's be come a META issue.

**Comment by KaelWD** [comment]
- VDataTableColumn is a component not a type, I have no idea how that is working. There isn't even a VDataTableColumn.d.ts
- vuetify/lib is not supported
- DataTableHeader and other types are still not exported

**Comment by npbenjohnson** [comment]
Workaround tested in ts 5.1 but should work in some older versions:
import type { VDataTable } from "vuetify/labs/VDataTable";
// For array props
type VDataTableHeader = Exclude<NonNullable<VDataTable["$props"]["headers"]>[number], Readonly<unknown[]>>;
// For non-array props
type SortItem = NonNullable<VDataTable["$props"]["sortBy"]>[number];

**Comment by SergioFerrando** [comment]
DataTableHeader and other types are still not exported. Any solution for this??  : (

**Comment by MatthewAry** [comment]
@KaelWD Perhaps a solution to making component prop types accessible could be found at [packages/component-meta](https://github.com/vuejs/language-tools/tree/master/packages/component-meta#vue-component-meta)

**Comment by MatthewAry** [comment]
I found a partial solution.
```typescript
import VSheet from 'vuetify/components/VSheet'
import vue from 'vue'

type VSheetPropTypes = Partial<vue.ExtractPropTypes<typeof VSheet>['VSheet']>
```
I say partial because while you can get all of the prop types using this trick, there is a bunch of other unnecessary stuff in the types that get resolved that I can't seem to use `Omit` to remove, including:

- `$`
- `$attrs`
- `$children`
- `$data`
- `$el`
- `$emit`
- `$forceUpdate`
- `$nextTick`
- `$options`
- `$parent`
- `$props`
- `$refs`
- `$root`
- `$slots`
- `$watch`
- `__isFragment`
- `__isSuspense`
- `__isTeleport`

**Comment by MatthewAry** [comment]
ValidationRule should also get exported. https://github.com/vuetifyjs/vuetify/blob/b87b28f72990ca0227906a266bbb58793332fcb8/packages/vuetify/src/composables/validation.ts#L15C1-L20C52

**Comment by MatthewAry** [comment]
This doesn't seem to work with Vuetify Component Types either. https://github.com/vuejs/language-tools/tree/master/packages/component-type-helpers

**Comment by volarname** [comment]
i asbolutely aggree that vuetify needs more type/interfaces export. probably not whole component props but complex types used in props like `DataTableHeader`, `DataTableSelectStrategy` and similar
of course it can be done by 3rd party @types library where types will be copy/pasted from vuetify library but as vuetify 3 is written completely in typescript and its just about the build scripts of existing types the best way will be to officially support it and exort in vuetify d.ts files

**Comment by MatthewAry** [comment]
For reference there is this conversation happening on Discord about extending components and passing types up. https://discord.com/channels/340160225338195969/1192558069343854612

**Comment by throrin19** [comment]
> This is very specifically about Labs, right? There is currently nothing to export anything other than the components.
> 
> Edit: I see the `Anchor` issue. Looks like this issue is being purposed for all of Vuetify

Any news about the Anchor part ? If we made a props used in location props of VMenu or VTooltip, we can't use it because it must be `Anchor | undefined` but we can't import Anchor....

**Comment by lenargum** [comment]
> > This is very specifically about Labs, right? There is currently nothing to export anything other than the components.
> > Edit: I see the `Anchor` issue. Looks like this issue is being purposed for all of Vuetify
> 
> Any news about the Anchor part ? If we made a props used in location props of VMenu or VTooltip, we can't use it because it must be `Anchor | undefined` but we can't import Anchor....

Here is dirty workaround if anyone interested. Just declare `Anchor` type manually 😁

```
declare const block: readonly ["top", "bottom"];
declare const inline: readonly ["start", "end", "left", "right"];
type Tblock = typeof block[number];
type Tinline = typeof inline[number];
type Anchor = Tblock | Tinline | 'center' | 'center center' | `${Tblock} ${Tinline | 'center'}` | `${Tinline} ${Tblock | 'center'}`;
```

**Comment by throrin19** [comment]
@lenargum this is not a long term solution. For that vuetify team must export it.

**Comment by mcfarljw** [comment]
I find myself wanting to get the exported interface shape of emitted events frequently. I don't see it mentioned specifically here so I'm posting in my use case with needing the typings from emitted events for reference:

```ts
// Vuetify
export interface SomeVuetifyEmitEvent { 
    id: unknown
    value: boolean
    path: unknown[]
    event: MouseEvent | PointerEvent
}

emits: {
    'update:selected': (value: unknown[]) => true,
    'update:opened': (value: unknown[]) => true,
    'click:open': (value: SomeVuetifyEmitEvent) => true,
    'click:select': (value: SomeVuetifyEmitEvent) => true,
},

// Client
import type { SomeVuetifyEmitEvent } from 'vuetify/VSelect'

function onSortSelect(event: SomeVuetifyEmitEvent) {
    const id = event.id as string
    const paths event.paths as string[]
}
```

**Comment by MatthewAry** [comment]
Looking to get prop types for DataTables?

```ts
type VDataTableProps = InstanceType<typeof VDataTable>['$props']
```
<img width="597" alt="Screenshot 2024-03-26 at 6 38 02 PM" src="https://github.com/vuetifyjs/vuetify/assets/157217/26b4c521-e82f-49c9-9843-aaa0a5b7e09d">
 
This might help until the promised day of better types comes.

**Comment by msladek** [comment]
How do I get at generic types like `DataTableItem<T>`, `ItemSlotBase<T>` and `ItemKeySlot<T>` needed for [cell-props](https://vuetifyjs.com/en/api/v-data-table/#props-cell-props) or [row-props](https://vuetifyjs.com/en/api/v-data-table/#props-row-props) as a workaround?

**Comment by jscottsf** [comment]
In case anyone is trying to do this with JSDoc -- this works.

```vue
    <v-data-table
      v-model:items-per-page="table.itemsPerPage"
      v-model:sort-by="table.sortBy"
      :headers="table.headers"
      :loading="loading"
      :items="scores"
      :items-per-page-options="table.itemsPerPageOptions"
    >
...
```

```vue
<script>
/**
 * @typedef { import("vuetify/components").VDataTable["$props"]["headers"] } VDataTableHeaders
 * @typedef { import("vuetify/components").VDataTable["$props"]["sortBy"] } VDataTableSortItems
 */
</script>
```

```vue
<script setup>
const table = reactive({
  /** @type {VDataTableHeaders} */
  headers: [
    { sortable: true, title: 'Label', value: 'label' },
    { align: 'end', sortable: true, title: 'Score', value: 'score' }
  ],
  itemsPerPage: 5,
  itemsPerPageOptions: [
    { value: 5, title: '5' },
    { value: -1, title: '$vuetify.dataFooter.itemsPerPageAll' }
  ],
  /** @type {VDataTableSortItems} */
  sortBy: [{ key: 'score', order: 'desc' }]
})
...
```

**Comment by ramseyfeng** [comment]
Any update on this issue?

**Comment by davidstackio** [comment]
To get my headers working without the typescript error I used `as const` as suggested here https://github.com/vuetifyjs/vuetify/issues/19599.

Example:

```ts
<script setup lang="ts">
const headers = [
  {
    title: "Dessert (100g serving)",
    align: "start",
    sortable: false,
    key: "name",
  },
  { title: "Calories", key: "calories", align: "end" },
  { title: "Fat (g)", key: "fat", align: "end" },
  { title: "Carbs (g)", key: "carbs", align: "end" },
  { title: "Protein (g)", key: "protein", align: "end" },
  { title: "Iron (%)", key: "iron", align: "end" },
] as const;
</script>
```

**Comment by gabrieljbaker** [comment]
seconding the importance of this 

I get TS errors when importing:

import {
  LocationStrategyData,
  StrategyProps,
} from "vuetify/lib/components/VOverlay/locationStrategies";

**Comment by darkbasic** [comment]
I second this: vuetify lacks important features like resizable tables which can only be implemented via dynamic wrapper components which unfortunately will get you rid of any typings for the props/slots/etc because you cannot import them from vuetify.

**Comment by septatrix** [comment]
Bonus points when also adding generics support: https://github.com/vuejs/core/pull/7963. That would be ideal for stuff like data tables etc

**Comment by Ricky816845** [comment]
It's been a year，Any updates?

**Comment by KaelWD** [comment]
#20819 allows anything to be imported from `vuetify/lib/whatever.js`. I'm going to leave this issue open for now though as internal types are sometimes moved to different files so importing from /lib is exposing implementation details, we still need to decide which interfaces need to be public and re-export them from either `vuetify`, `vuetify/components`, or `vuetify/components/<NAME>`.

**Comment by Scandal-UK** [comment]
Regarding `Default export of the module has or is using private name 'SubmitEventPromise'`, are there any known workarounds to prevent this same old error when adding VForm to our project?

Seems to me we must either downgrade vuetify or typescript to compile templates - or remove v-form instances, I suppose

**Comment by jannes-io** [comment]
Same for `VAutoComplete`, would appreciate exports on all of the `Filter*` interfaces.

**Comment by RonAruch** [comment]
Do you happen to have any updates about this feature?
Access to the component props type is a basic need.
